# <div style="text-align: center">Advanced Python Project - Vin de Renom</div>
_<div style="text-align: center">Nicolas POUGET, Mohammad AWWAD, Mohamad ALHALABI</div>_

In [1]:
import pandas as pd
import plotly.express as px
import numpy as np

## _Partie 1: Exploration_

### _On commence par importer les données et examiner les colonnes :_

In [2]:
cms_df = pd.read_excel('web.xlsx')
cms_df

D:\Users\mOh\anaconda3\lib\site-packages\openpyxl\worksheet\_reader.py:312: UserWarning: Unknown extension is not supported and will be removed
  warn(msg)


sku  virtual  downloadable  rating_count  \
0     bon-cadeau-25-euros        0             0             0   
1                   15298        0             0             0   
2                   15296        0             0             0   
3                   15300        0             0             0   
4                   19814        0             0             0   
...                   ...      ...           ...           ...   
1508                16135        0             0             0   
1509                15891        0             0             0   
1510                15887        0             0             0   
1511              13127-1        0             0             0   
1512                16230        0             0             0   

      average_rating  total_sales tax_status  tax_class  post_author  \
0                0.0         10.0    taxable        NaN          1.0   
1                0.0          6.0    taxable        NaN          2.0   
2                0.0          0.0    taxable        NaN          2.0   
3                0.0          0.0    taxable        NaN          2.0   
4                0.0          3.0    taxable        NaN          2.0   
...              ...          ...        ...        ...          ...   
1508             0.0          5.0        NaN        NaN          2.0   
1509             0.0          0.0        NaN        NaN          2.0   
1510             0.0          0.0        NaN        NaN          2.0   
1511             0.0          0.0        NaN        NaN          2.0   
1512             0.0          0.0        NaN        NaN          2.0   

               post_date  ...  \
0    2018-06-01 13:53:46  ...   
1    2018-02-08 12:58:52  ...   
2    2018-02-08 13:49:41  ...   
3    2018-02-08 14:08:36  ...   
4    2018-02-09 14:01:05  ...   
...                  ...  ...   
1508 2020-04-25 13:22:38  ...   
1509 2020-05-02 14:53:40  ...   
1510 2020-05-02 15:00:54  ...   
1511 2020-06-09 15:42:04  ...   
1512 2020-07-20 11:00:00  ...   

                                              post_name       post_modified  \
0                                bon-cadeau-de-25-euros 2018-06-01 14:13:57   
1           pierre-jean-villa-saint-joseph-preface-2018 2019-12-30 09:30:29   
2             pierre-jean-villa-saint-joseph-tilde-2017 2019-12-21 09:00:17   
3     pierre-jean-villa-croze-hermitage-accroche-coe... 2020-06-26 18:15:03   
4                     pierre-jean-villa-igp-gamine-2018 2020-01-04 16:36:01   
...                                                 ...                 ...   
1508                    mouthes-le-bihan-aime-chai-2015 2020-08-26 17:35:03   
1509         camin-larredya-jurancon-sec-la-virada-2018 2020-08-26 17:35:02   
1510             jamet-cote-rotie-fructus-voluptas-2018 2020-08-14 18:15:03   
1511     clos-du-mont-olivet-chateauneuf-du-pape-2007-2 2020-07-20 17:09:06   
1512  domaine-saint-nicolas-fiefs-vendeens-blanc-les... 2020-08-13 10:45:03   

       post_modified_gmt post_content_filtered post_parent  \
0    2018-06-01 12:13:57                   NaN         0.0   
1    2019-12-30 08:30:29                   NaN         0.0   
2    2019-12-21 08:00:17                   NaN         0.0   
3    2020-06-26 16:15:03                   NaN         0.0   
4    2020-01-04 15:36:01                   NaN         0.0   
...                  ...                   ...         ...   
1508 2020-08-26 15:35:03                   NaN         0.0   
1509 2020-08-26 15:35:02                   NaN         0.0   
1510 2020-08-14 16:15:03                   NaN         0.0   
1511 2020-07-20 15:09:06                   NaN         0.0   
1512 2020-08-13 08:45:03                   NaN         0.0   

                                                   guid menu_order  \
0     https://www.wine-spirit.fr/?post_type=product&...        0.0   
1     https://www.wine-spirit.fr/?post_type=product&...        0.0   
2     https://www.wine-spirit.fr/?post_type=product&...        

In [3]:
cms_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1513 entries, 0 to 1512
Data columns (total 28 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   sku                    1428 non-null   object        
 1   virtual                1513 non-null   int64         
 2   downloadable           1513 non-null   int64         
 3   rating_count           1513 non-null   int64         
 4   average_rating         1430 non-null   float64       
 5   total_sales            1430 non-null   float64       
 6   tax_status             716 non-null    object        
 7   tax_class              0 non-null      float64       
 8   post_author            1430 non-null   float64       
 9   post_date              1430 non-null   datetime64[ns]
 10  post_date_gmt          1430 non-null   datetime64[ns]
 11  post_content           0 non-null      float64       
 12  post_title             1430 non-null   object        
 13  pos

### _On voit que la colonne `sku` n'est pas totalement non null comme elle doit l'être. Elle doit representer les `order_id` qui sert à identifier de facon unique chaque transaction._

In [4]:
erp_df = pd.read_excel('erp.xlsx')
erp_df

D:\Users\mOh\anaconda3\lib\site-packages\openpyxl\worksheet\_reader.py:312: UserWarning: Unknown extension is not supported and will be removed
  warn(msg)


product_id  onsale_web  price  stock_quantity stock_status
0          3847           1   24.2               0   outofstock
1          3849           1   34.3               0   outofstock
2          3850           1   20.8               0   outofstock
3          4032           1   14.1               0   outofstock
4          4039           1   46.0               0   outofstock
..          ...         ...    ...             ...          ...
820        7203           0   45.0              30      instock
821        7204           0   45.0               9      instock
822        7247           1   54.8              23      instock
823        7329           0   26.5              14      instock
824        7338           1   16.3              45      instock

[825 rows x 5 columns]

In [5]:
erp_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 825 entries, 0 to 824
Data columns (total 5 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   product_id      825 non-null    int64  
 1   onsale_web      825 non-null    int64  
 2   price           825 non-null    float64
 3   stock_quantity  825 non-null    int64  
 4   stock_status    825 non-null    object 
dtypes: float64(1), int64(3), object(1)
memory usage: 32.4+ KB


### _Pas de colonnes vides dans cette table, mais une relation existe entre les deux dernières colonnes et donc il faut vérifier sa validité_

In [6]:
key_df = pd.read_excel('liaison.xlsx')
key_df

D:\Users\mOh\anaconda3\lib\site-packages\openpyxl\worksheet\_reader.py:312: UserWarning: Unknown extension is not supported and will be removed
  warn(msg)


product_id   id_web
0          3847    15298
1          3849    15296
2          3850    15300
3          4032    19814
4          4039    19815
..          ...      ...
820        7203      NaN
821        7204      NaN
822        7247  13127-1
823        7329  14680-1
824        7338    16230

[825 rows x 2 columns]

In [7]:
key_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 825 entries, 0 to 824
Data columns (total 2 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   product_id  825 non-null    int64 
 1   id_web      734 non-null    object
dtypes: int64(1), object(1)
memory usage: 13.0+ KB


### _On peut voir que la table des liaisons manquent des valeurs (734 non-null / 825 total). Etant une table de liaison, il faut prévoir une relation One-to-One dans cette table et donc on ne peut pas laisser des valeurs vides_

## _Partie 2 : Nettoyage_

### _On commence par la table du CMS pour voir les lignes sans `sku` (null) et la possibilité de les éliminer_

In [8]:
# Filtrer les lignes ou le SKU est NaN
cms_df[cms_df.sku.isna()]

sku  virtual  downloadable  rating_count  average_rating  total_sales  \
178  NaN        0             0             0             NaN          NaN   
179  NaN        0             0             0             NaN          NaN   
227  NaN        0             0             0             NaN          NaN   
230  NaN        0             0             0             NaN          NaN   
231  NaN        0             0             0             NaN          NaN   
..   ...      ...           ...           ...             ...          ...   
792  NaN        0             0             0             NaN          NaN   
793  NaN        0             0             0             NaN          NaN   
794  NaN        0             0             0             NaN          NaN   
795  NaN        0             0             0             NaN          NaN   
796  NaN        0             0             0             NaN          NaN   

    tax_status  tax_class  post_author post_date  ... post_name  \
178        NaN        NaN          NaN       NaT  ...       NaN   
179        NaN        NaN          NaN       NaT  ...       NaN   
227        NaN        NaN          NaN       NaT  ...       NaN   
230        NaN        NaN          NaN       NaT  ...       NaN   
231        NaN        NaN          NaN       NaT  ...       NaN   
..         ...        ...          ...       ...  ...       ...   
792        NaN        NaN          NaN       NaT  ...       NaN   
793        NaN        NaN          NaN       NaT  ...       NaN   
794        NaN        NaN          NaN       NaT  ...       NaN   
795        NaN        NaN          NaN       NaT  ...       NaN   
796        NaN        NaN          NaN       NaT  ...       NaN   

     post_modified post_modified_gmt post_content_filtered post_parent guid  \
178            NaT               NaT                   NaN         NaN  NaN   
179            NaT               NaT                   NaN         NaN  NaN   
227            NaT               NaT                   NaN         NaN  NaN   
230            NaT               NaT                   NaN         NaN  NaN   
231            NaT               NaT                   NaN         NaN  NaN   
..             ...               ...                   ...         ...  ...   
792            NaT               NaT                   NaN         NaN  NaN   
793            NaT               NaT                   NaN         NaN  NaN   
794            NaT               NaT                   NaN         NaN  NaN   
795            NaT               NaT                   NaN         NaN  NaN   
796            NaT               NaT                   NaN         NaN  NaN   

    menu_order  post_type post_mime_type comment_count  
178        NaN        NaN            NaN           NaN  
179        NaN        NaN            NaN           NaN  
227        NaN        NaN            NaN           NaN  
230        NaN        NaN            NaN           NaN  
231        NaN        NaN            NaN           NaN  
..         ...        ...            ...           ...  
792        NaN        NaN            NaN           NaN  
793        NaN        NaN            NaN           NaN  
794        NaN        NaN            NaN           NaN  
795        NaN        NaN            NaN           NaN  
796        NaN        NaN            NaN           NaN  

[85 rows x 28 columns]

### _D'un aperçu rapide, il semble que ces lignes sont bien inutiles mais on check davantage pour avoir le feu vert._

In [9]:
cms_df[cms_df.sku.isna()].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 85 entries, 178 to 796
Data columns (total 28 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   sku                    0 non-null      object        
 1   virtual                85 non-null     int64         
 2   downloadable           85 non-null     int64         
 3   rating_count           85 non-null     int64         
 4   average_rating         2 non-null      float64       
 5   total_sales            2 non-null      float64       
 6   tax_status             2 non-null      object        
 7   tax_class              0 non-null      float64       
 8   post_author            2 non-null      float64       
 9   post_date              2 non-null      datetime64[ns]
 10  post_date_gmt          2 non-null      datetime64[ns]
 11  post_content           0 non-null      float64       
 12  post_title             2 non-null      object        
 13  post

### _Il semble que certaines lignes ne sont pas vides, il faut filter pour les voir._

In [10]:
cms_df[cms_df.sku.isna()][-cms_df[cms_df.sku.isna()].total_sales.isna()]


sku  virtual  downloadable  rating_count  average_rating  total_sales  \
470  NaN        0             0             0             0.0          0.0   
471  NaN        0             0             0             0.0          0.0   

    tax_status  tax_class  post_author           post_date  ...  \
470    taxable        NaN          2.0 2018-07-31 12:07:23  ...   
471    taxable        NaN          2.0 2018-08-08 11:23:43  ...   

                                      post_name       post_modified  \
470  pierre-jean-villa-cote-rotie-fongeant-2017 2019-11-02 13:24:15   
471    pierre-jean-villa-condrieu-suspendu-2018 2019-11-02 13:24:01   

      post_modified_gmt post_content_filtered post_parent  \
470 2019-11-02 12:24:15                   NaN         0.0   
471 2019-11-02 12:24:01                   NaN         0.0   

                                                  guid menu_order  post_type  \
470  https://www.wine-spirit.fr/?post_type=product&...        0.0    product   
471  https://www.wine-spirit.fr/?post_type=product&...        0.0    product   

    post_mime_type comment_count  
470            NaN           0.0  
471            NaN           0.0  

[2 rows x 28 columns]

### _Vu que `total_sales` est 0 pour ces deux lignes, on peut s'en debarrasser._

In [11]:
cleaned_cms_df = cms_df.dropna(subset=['sku'])
cleaned_cms_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1428 entries, 0 to 1512
Data columns (total 28 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   sku                    1428 non-null   object        
 1   virtual                1428 non-null   int64         
 2   downloadable           1428 non-null   int64         
 3   rating_count           1428 non-null   int64         
 4   average_rating         1428 non-null   float64       
 5   total_sales            1428 non-null   float64       
 6   tax_status             714 non-null    object        
 7   tax_class              0 non-null      float64       
 8   post_author            1428 non-null   float64       
 9   post_date              1428 non-null   datetime64[ns]
 10  post_date_gmt          1428 non-null   datetime64[ns]
 11  post_content           0 non-null      float64       
 12  post_title             1428 non-null   object        
 13  pos

### _On peut également enlever tous les lignes ou `total_sales` est 0 vu que ca n'affecte pas le chiffre d'affaire et pollue le nombre de vrai ventes._

In [12]:
cleaned_cms_df.total_sales.describe()


count    1428.000000
mean        4.012605
std         8.515198
min         0.000000
25%         0.000000
50%         1.000000
75%         4.000000
max        96.000000
Name: total_sales, dtype: float64

### _On a un min de 0 dans `total_sales` actuellement._

In [13]:
cleaned_cms_df = cleaned_cms_df[cleaned_cms_df.total_sales != 0]
cleaned_cms_df.total_sales.describe()


count    770.000000
mean       7.441558
std       10.440361
min        1.000000
25%        2.000000
50%        4.000000
75%        9.000000
max       96.000000
Name: total_sales, dtype: float64

### _A partir du `post_type`, on a chaque produit en doublons donc on élimine les `attachments`_

In [14]:
cleaned_cms_df = cleaned_cms_df[cleaned_cms_df.post_type != 'attachment']

### _On passe ensuite à valider la table `erp`._ 

In [15]:
# Filtrer les lignes ou stock quantity est 0 mais le statut n'est pas out of stock
erp_df[erp_df.stock_quantity ==
       0][erp_df[erp_df.stock_quantity == 0].stock_status != 'outofstock']


product_id  onsale_web  price  stock_quantity stock_status
443        4954           1   25.0               0      instock

### _On corrige cette ligne_

In [16]:
cleaned_erp_df = erp_df
cleaned_erp_df.at[443, 'stock_status'] = 'outofstock'
cleaned_erp_df[cleaned_erp_df.stock_quantity ==
               0][cleaned_erp_df[cleaned_erp_df.stock_quantity == 0].stock_status != 'outofstock']


Empty DataFrame
Columns: [product_id, onsale_web, price, stock_quantity, stock_status]
Index: []

### _La colonne `onsale_web` contient des nulles qu'on souhaite corriger_

In [17]:
cleaned_erp_df['onsale_web'] = cleaned_erp_df.onsale_web.apply(lambda x: 1 if x == 1 else 0)
cleaned_erp_df

product_id  onsale_web  price  stock_quantity stock_status
0          3847           1   24.2               0   outofstock
1          3849           1   34.3               0   outofstock
2          3850           1   20.8               0   outofstock
3          4032           1   14.1               0   outofstock
4          4039           1   46.0               0   outofstock
..          ...         ...    ...             ...          ...
820        7203           0   45.0              30      instock
821        7204           0   45.0               9      instock
822        7247           1   54.8              23      instock
823        7329           0   26.5              14      instock
824        7338           1   16.3              45      instock

[825 rows x 5 columns]

### _Mainteant, à l'aide de la table `liaison`, on check si les produits n'ayant pas un `web_id`/`sku` ont été référencés dans `cms`_

In [18]:
# JOIN ERP WITH KEY (LIAISON) TO ADD COL ID_WEB/SKU
merged_erp = cleaned_erp_df.merge(key_df, on='product_id')
merged_erp.dropna(subset=['id_web'], inplace=True)
merged_erp

product_id  onsale_web  price  stock_quantity stock_status   id_web
0          3847           1   24.2               0   outofstock    15298
1          3849           1   34.3               0   outofstock    15296
2          3850           1   20.8               0   outofstock    15300
3          4032           1   14.1               0   outofstock    19814
4          4039           1   46.0               0   outofstock    19815
..          ...         ...    ...             ...          ...      ...
792        7023           1   27.5              15      instock    15891
793        7025           1   69.0               2      instock    15887
822        7247           1   54.8              23      instock  13127-1
823        7329           0   26.5              14      instock  14680-1
824        7338           1   16.3              45      instock    16230

[734 rows x 6 columns]

In [19]:
# LEFT JOIN CMS WITH ERP 
centered_df = cleaned_cms_df.merge(merged_erp, left_on='sku', right_on='id_web')
centered_df

sku  virtual  downloadable  rating_count  average_rating  \
0    bon-cadeau-25-euros        0             0             0             0.0   
1                  15298        0             0             0             0.0   
2                  19814        0             0             0             0.0   
3                  16042        0             0             0             0.0   
4                  14980        0             0             0             0.0   
..                   ...      ...           ...           ...             ...   
380                16068        0             0             0             0.0   
381                 8193        0             0             0             0.0   
382                15845        0             0             0             0.0   
383                15741        0             0             0             0.0   
384                16135        0             0             0             0.0   

     total_sales tax_status  tax_class  post_author           post_date  ...  \
0           10.0    taxable        NaN          1.0 2018-06-01 13:53:46  ...   
1            6.0    taxable        NaN          2.0 2018-02-08 12:58:52  ...   
2            3.0    taxable        NaN          2.0 2018-02-09 14:01:05  ...   
3            7.0    taxable        NaN          2.0 2018-02-12 09:54:27  ...   
4            3.0    taxable        NaN          2.0 2018-02-12 10:03:05  ...   
..           ...        ...        ...          ...                 ...  ...   
380          1.0    taxable        NaN          2.0 2020-02-22 14:28:32  ...   
381          3.0    taxable        NaN          2.0 2020-03-04 13:54:12  ...   
382          1.0    taxable        NaN          2.0 2020-04-25 12:43:23  ...   
383          2.0    taxable        NaN          2.0 2020-04-25 12:49:49  ...   
384          5.0    taxable        NaN          2.0 2020-04-25 13:22:38  ...   

    menu_order  post_type post_mime_type comment_count product_id onsale_web  \
0          0.0    product            NaN           0.0       4954          1   
1          0.0    product            NaN           0.0       3847          1   
2          0.0    product            NaN           0.0       4032          1   
3          0.0    product            NaN           0.0       4042          1   
4          0.0    product            NaN           0.0       4043          1   
..         ...        ...            ...           ...        ...        ...   
380        0.0    product            NaN           0.0       6665          1   
381        0.0    product            NaN           0.0       6738          1   
382        0.0    product            NaN           0.0       6926          1   
383        0.0    product            NaN           0.0       6928          1   
384        0.0    product            NaN           0.0       6930          1   

    price  stock_quantity stock_status               id_web  
0    25.0               0   outofstock  bon-cadeau-25-euros  
1    24.2               0   outofstock                15298  
2    14.1               0   outofstock                19814  
3    31.2              34      instock                16042  
4    60.0              12      instock                14980  
..    ...             ...          ...                  ...  
380  27.7               8      instock                16068  
381  15.4              40      instock                 8193  
382  49.9              24      instock                15845  
383  19.0              20      instock                15741  
384   8.4              83      instock                16135  

[385 rows x 34 columns]

### _Après avoir rapprocher les tables, on voit si certain `sku` n'ont pas de `product_id`_

In [20]:
centered_df[centered_df.product_id.isna()]

Empty DataFrame
Columns: [sku, virtual, downloadable, rating_count, average_rating, total_sales, tax_status, tax_class, post_author, post_date, post_date_gmt, post_content, post_title, post_excerpt, post_status, comment_status, ping_status, post_password, post_name, post_modified, post_modified_gmt, post_content_filtered, post_parent, guid, menu_order, post_type, post_mime_type, comment_count, product_id, onsale_web, price, stock_quantity, stock_status, id_web]
Index: []

[0 rows x 34 columns]

### _Pas de lignes n'ayant de `product_id`._

## _Partie 3 :  Analyse_

### _Pour trouver le chiffre d'affaire réalisé par produit, on fait :_

In [21]:
# On groupe par nom de produit et le prix et fait la somme des total_sales puis on les rend des colonnes pour pouvoir les utiliser
price_per_product = centered_df[['post_title', 'price', 'total_sales']].groupby(
    ['post_title', 'price']).sum().reset_index(level=['post_title', 'price'])
# CA = Qtt x Prix
price_per_product['chiffre_affaire'] = price_per_product.price * \
    price_per_product.total_sales
price_per_product


post_title  price  total_sales  \
0                Agnès Levet Côte Rôtie Améthyste 2017   41.2          4.0   
1                 Agnès Levet Côte Rôtie Maestria 2017   55.0          3.0   
2                 Agnès Levet Côte Rôtie Péroline 2017   57.0          3.0   
3           Alain Graillot Crozes-Hermitage Blanc 2018   24.4         10.0   
4           Alain Graillot Crozes-Hermitage Rouge 2017   24.4          6.0   
..                                                 ...    ...          ...   
380    Xavier Frissant Touraine Les Roses du Clos 2018   12.8         15.0   
381            Xavier Frissant Touraine Sauvignon 2019    9.7          3.0   
382               Zind-Humbrecht Muscat Turckheim 2016   17.9          1.0   
383  Zind-Humbrecht Riesling Herrenweg de Turckheim...   31.7          2.0   
384                           Zind-Humbrecht Zind 2017   25.7          6.0   

     chiffre_affaire  
0              164.8  
1              165.0  
2              171.0  
3              244.0  
4              146.4  
..               ...  
380            192.0  
381             29.1  
382             17.9  
383             63.4  
384            154.2  

[385 rows x 4 columns]

### _Le chiffre d'affaire total est :_

In [22]:
price_per_product['chiffre_affaire'].sum()

70568.6

### _Concernant les prix des produits:_

In [23]:
price_per_product.price.describe()

count    385.000000
mean      26.581429
std       24.934530
min        5.700000
25%       12.500000
50%       19.000000
75%       31.700000
max      225.000000
Name: price, dtype: float64

### _Pour visualiser des outliers dans les prix des produits, on commence par la méthodes des interquartiles:_

### _Méthode graphique_

### _Boxplot_

In [24]:
px.box(price_per_product, y="price", height=1500, title='Distribution des prix par produit')

### _On a beaucoup de outliers qui sont situés au-delà du 1.5IQR + Q3 dans les prix établis_ 

In [25]:
px.box(price_per_product, y="total_sales", height=1500, title='Distribution des ventes par produit')

### _Certaines variables n'ont pas d'influence sur les prix, donc on peut les éliminer_

In [26]:
reduced_df = centered_df.drop(columns=['virtual', 'downloadable', 'rating_count', 'average_rating', 'tax_class', 'post_date', 'menu_order', 'post_mime_type', 'comment_count', 'post_date_gmt',
                              'post_content',	'post_excerpt',	'post_status',	'comment_status',	'ping_status',	'post_password',	'post_name',	'post_modified',	'post_modified_gmt',	'post_content_filtered',	'post_parent',	'guid', 'tax_status', 'post_type', 'id_web', 'post_author', 'post_title', 'onsale_web'])
reduced_df.head()

sku  total_sales  product_id  price  stock_quantity  \
0  bon-cadeau-25-euros         10.0        4954   25.0               0   
1                15298          6.0        3847   24.2               0   
2                19814          3.0        4032   14.1               0   
3                16042          7.0        4042   31.2              34   
4                14980          3.0        4043   60.0              12   

  stock_status  
0   outofstock  
1   outofstock  
2   outofstock  
3      instock  
4      instock

### _Calcul statistique_

### _En utilisant le calcul du `IQR = Q3 - Q1`, on détecte les outliers au-dessus de `1.5IQR + Q3`_

In [27]:
iqr_price = reduced_df.price.quantile(0.75) - reduced_df.price.quantile(0.25)
iqr_sales = reduced_df.total_sales.quantile(0.75) - reduced_df.total_sales.quantile(0.25)
conditions = [
    reduced_df.price > reduced_df.price.quantile(0.75) + 1.5 * iqr_price,
    reduced_df.total_sales > reduced_df.total_sales.quantile(0.75) + 1.5 * iqr_sales
]
choices = ['Price_Outlier','Sale_Outliers']
reduced_df['outliers'] = np.select(conditions, choices, default='Not_Outlier')
px.scatter(reduced_df, x='price', y='total_sales', height=1500, title='Distribution des prix par produit', color='outliers')

### _Z score_

### _On peut calculer le Z-score par Z = X - avg / stddev_

In [34]:
reduced_df['z_score_price'] = reduced_df.price.apply(lambda x: (
    x - reduced_df.price.mean())/reduced_df.price.std(ddof=0))
reduced_df['z_score_sales'] = reduced_df.total_sales.apply(lambda x: (
    x - reduced_df.total_sales.mean()))/reduced_df.total_sales.std(ddof=0)
# Si z_score > 2*stddev -> Outlier
reduced_df[reduced_df['z_score_sales'].abs() > 2]

sku  total_sales  product_id  price  stock_quantity stock_status  \
24   16416         62.0        4068   16.6             157      instock   
45     304         40.0        4141   39.0               1      instock   
46   11641         30.0        4142   53.0               8      instock   
47    1662         87.0        4144   49.0              11      instock   
52   16237         36.0        4153   29.0               0   outofstock   
67   16210         43.0        4172    5.7             167      instock   
76   16189         42.0        4187   13.3              90      instock   
85   16295         46.0        4200    5.8             190      instock   
111  16317         30.0        4250   19.5              14      instock   
134   7818         96.0        4334   49.0               0   outofstock   
190  15349         32.0        4706   16.8              23      instock   
205     38         38.0        4729    8.6             151      instock   
223  15307         29.0        4861    8.5             284      instock   
228  16149         33.0        4870    9.3               0   outofstock   
232  15807         36.0        4891   27.9               0   outofstock   
352  16264         38.0        6047   10.9              46      instock   
363  16580         41.0        6206   25.2             120      instock   
364  16077         37.0        6207   25.2             363      instock   

          outliers  z_score_price  z_score_sales  
24   Sale_Outliers      -0.400826       5.229120  
45   Sale_Outliers       0.498695       3.120544  
46   Sale_Outliers       1.060896       2.162100  
47   Sale_Outliers       0.900267       7.625230  
52   Sale_Outliers       0.097123       2.737166  
67   Sale_Outliers      -0.838540       3.408077  
76   Sale_Outliers      -0.533345       3.312233  
85   Sale_Outliers      -0.834524       3.695610  
111  Sale_Outliers      -0.284370       2.162100  
134  Sale_Outliers       0.900267       8.487830  
190  Sale_Outliers      -0.392795       2.353789  
205  Sale_Outliers      -0.722084       2.928855  
223  Sale_Outliers      -0.726100       2.066256  
228  Sale_Outliers      -0.693974       2.449633  
232  Sale_Outliers       0.052950       2.737166  
352  Sale_Outliers      -0.629722       2.928855  
363  Sale_Outliers      -0.055474       3.216388  
364  Sale_Outliers      -0.055474       2.833011

In [35]:
reduced_df[reduced_df['z_score_price'].abs() > 2]

sku  total_sales  product_id  price  stock_quantity stock_status  \
6    15269          6.0        4046   80.0               0   outofstock   
136  15940          5.0        4352  225.0               0   outofstock   
137  12587          3.0        4353   79.5              11      instock   
138  12589         11.0        4355  126.5               2      instock   
140  13853          1.0        4359   85.6               0   outofstock   
144   3510         13.0        4402  176.0               8      instock   
145   3507          2.0        4404  108.5               2      instock   
146   3509          1.0        4407  104.0               6      instock   
237  14220          5.0        4904  137.0              13      instock   
339  14983          3.0        5892  191.3              10      instock   
359  14923          2.0        6126  135.0              10      instock   
365  13996          2.0        6212  115.0               2      instock   

          outliers  z_score_price  z_score_sales  
6    Price_Outlier       2.145141      -0.138165  
136  Price_Outlier       7.967937      -0.234010  
137  Price_Outlier       2.125062      -0.425698  
138  Price_Outlier       4.012451       0.341057  
140  Price_Outlier       2.370021      -0.617387  
144  Price_Outlier       6.000234       0.532745  
145  Price_Outlier       3.289622      -0.521543  
146  Price_Outlier       3.108914      -0.617387  
237  Price_Outlier       4.434102      -0.234010  
339  Price_Outlier       6.614639      -0.425698  
359  Price_Outlier       4.353788      -0.521543  
365  Price_Outlier       3.550643      -0.521543